src/main/scala/is/hail/expr/AST.scala

```diff
@@ -691,6 +724,25 @@ case class Apply(posn: Position, fn: String, args: Array[AST]) extends AST(posn,
     case (_, _) =>
       FunctionRegistry.call(fn, args, args.map(_.`type`).toSeq)
   }
+
+  private val tryPrimOpConversion: IndexedSeq[IR] => Option[IR] = {
+    val pf: PartialFunction[IndexedSeq[IR], Option[IR]] = {
+      case IndexedSeq(x) => for {
+        op <- ir.UnaryOp.fromString.lift(fn)
+        t <- ir.UnaryOp.maybeReturnType(op, x.typ)
+      } yield ir.ApplyUnaryPrimOp(op, x, t)
+      case IndexedSeq(x, y) => for {
+        op <- ir.BinaryOp.fromString.lift(fn)
+        t <- ir.BinaryOp.maybeReturnType(op, x.typ, y.typ)
+      } yield ir.ApplyBinaryPrimOp(op, x, y, t)
+    }
+    pf.lift.andThen(_.flatten)
```

Comment:
Something about this offends me.  It feels like we're using two different mechanisms for missing/partialness that don't play well with each other.  What about defining a `PartialFunction[T, Option[U]].flatten: PartialFunction[T, U]`?  Although `flatten` doesn't quite sound right.  I don't know...

Reply:
Seems reasonable. It also bothered me but I also didn't have a good name for this operation.